Added To-Many relationship support when insert Parse object

callzhang opened the pull request:
Added support to To-Many relationship when creating PFObject in:
- (void)setValuesFromManagedObject:(NSManagedObject _)managedObject withSaveCallbacks:(NSMutableDictionary *_)saveCallbacks

Fixed a bug for save callback in:
- (void)addParseSaveCallbacks:(NSArray*)callbacks forObject:(NSManagedObjectID *)objectID

callzhang committed:
Added To-Many relationship support when insert Parse object

callzhang committed:
Updated PFIS to enable deletion:
1) Fixed a bug in to-many relation save callback
2) Added ‘isDeleted’ check before fetch for new value and relation
3) Check PFObject key existence before accessing its value

callzhang commented:
There is a small bug in the first submition in:
(void)addParseSaveCallbacks:(NSArray*)callbacks forObject:(NSManagedObjectID *)objectID

It is fixed in the second sumbittion.

Also added some control flow code to support the deletion on IncrementalStore. Still, it is very buggy and the process of deletion is not a clean cut, in which process the IncrementalStore tend to loop through deleted objects, and there is chances that strange exception could be thrown, even I tried to stop that intreating process as much as I can. 

I am developing my own app. Due to StackMob's shutdown, I am trying to find a solution for my local Core Data persistency. I need a higher level overview of the process to improve this framework. Currently, when I going through the code, a lot of things don't make sense to me, like the use of Child Context. It is discouraged to use child context though multiple articles I have read, and one can be found on Objc.io. Also, with backingContext off the main thread, why bother using a child context. 

I can see the potential of this framework, if you can get me more involved in this process, I would love to contribute more.

Alydyn commented:
@callzhang You committed your Parse application ID's.  You might want to remove those and re submit.

Alydyn commented on Examples/Basic Example/Basic Example/BERootViewController.m:
```diff
@@ -9,12 +9,15 @@
 
-@interface BERootViewController () <NSFetchedResultsControllerDelegate>
+@interface BERootViewController () <NSFetchedResultsControllerDelegate>{
+    NSManagedObjectContext *context;
```
This should probably be declared as a property instead.

Alydyn commented on Examples/Basic Example/Basic Example/BERootViewController.m:
```diff
@@ -145,8 +179,12 @@ - (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEd
         Artist *rowArtist = [[tableSection objects] objectAtIndex:indexPath.row];
-        [self.fetchedResultsController.managedObjectContext deleteObject:rowArtist];
-        [self.fetchedResultsController.managedObjectContext save:nil];
+        for (Song *s in rowArtist.artistSongs) {
```
`artistSongs` is defined with a cascade delete rule.  This bit of code to delete the related songs is unnecessary.

callzhang commented:
@Alydyn Thanks! With the cascade delete rule on, it generated a lot of problem(at beginning). So I just changed it to nullify, and delete manually to spottily the problem. So I can focus on the IncrementalStore. I think it will be the same.

callzhang commented:
@Alydyn This is the first time I use pull request. I need to learn how to submit only the part of the code I care about. The Parse ID is a test one, so it's fine to leave it there.

Alydyn commented:
@callzhang the Parse ID still needs to be removed, because if @sbonami were to accept this pull request then those app ID's would be in there for everyone, from that point on.  Although I do think he can accept the pull request selectively (not sure though, never tried it).

That is interesting regarding the cascade delete rule.  What was the problem you were seeing?

callzhang commented:
@Alydyn Sure. I will do another pull request. Do you know how to submit selectively?
The cascade delete should be the same as what I did manually. It is the looping of accessing relationship, and thus generating a lot of Parse SDK warning of "Object not found" (and occasionally some errors) , that made me to change cascade delete to nullify - an effort to isolate the issue. I will change it back to cascade delete and see if they are the same.

Alydyn commented:
@callzhang not sure once it has been committed.  Maybe @sbonami can give us an idea of what he would like to see?

callzhang committed:
Fixed a major bug in deleting: MO should be deleted no matter PFObject found or not.
This will stop the looping of accessing the deleted value and save a lot of network calls.
Remaining issue: The Artist->Songs relation could be lost sometime. Need further investigation.
